### ISSUE TITLE
Refactoring CFG file to separate setting from check and allow overriding by env var

### ISSUE THREAD

**Issue by inkhey** [tracim/tracim#1550]
**Currently**:
- No way to set config through env var in tracim
- Code for checking validity of variable and code for setting them is mixed.

Allow setting easily env var will make tests easier to custom.
CFG object will be cleaner.

**Expected**:
- Separated get value from config file/env var and check validity in CFG.
- add generic mecanism to override config through env var.
- Added: set var using get_raw_config() method, see https://github.com/tracim/tracim/issues/1550#issuecomment-480172784
- Added: better logging for config: source of config applied, etc...
- Rename all parameter with similar pattern: to upper and  "." -> "__" .
- Verify all parameters setted with get_raw_config exist as parameter of CFG object .

**Comment by inkhey** [comment]
@lebouquetin There is an issue with making env var mecanism generic
We have in entry always string but parameter should be something converted sometime to int, list or other type. We should know for all param how to convert them in both situation: when setting CFG from settings dict (from config file) and from env var value, currently we set to int explicitly for all parameters, so we can't make it generic for env var.

@buxx propose to change CFG object with a list of spec:
```
    available_configs = {
      'APP_ENABLED': ConfigProps(type_=ConfigList, setting_param='app.enabled', legacy_setting_param='app_enabled', env_var_name='TRACIM_APP_ENABLED'),
    }
```
type_ is callable to cast str setting,
setting_param is name of the param is settings dict
legacy_setting_param is to allow some legacy param for retrocompat with logging warning message,
env_var_name is just an example, we can this way decide to force a specific env_var name for parameter, not sure we need it (use convention like 'TRACIM_{}' prefixed env var is ok).

**Comment by lebouquetin** [comment]
could you give a concrete example of current config parameter which is a problem?

**Comment by inkhey** [comment]
@lebouquetin A simple example : 

in code we have : 
```
        self.EMAIL_REPLY_CONNECTION_MAX_LIFETIME = int(settings.get(
            'email.reply.connection.max_lifetime',
            600,  # 10 minutes
        ))
```
If we want to support env_var, we should do something like:
```
self.EMAIL_REPLY_CONNECTION_MAX_LIFETIME = int(os.env.get('TRACIM_EMAIL_REPLY_CONNECTION_MAX_LIFETIME'))
```
But if we want to automatically asign env_var according to name, in order to have a generic system, it's just imposible because we have no way to know that self.EMAIL_REPLY_CONNECTION_MAX_LIFETIME should be cast as int but not self.EMAIL_REPLY_LOCKFILE_PATH

**Comment by lebouquetin** [comment]
One solution would be to have the following design:

```
class RawConfig
  def __init__(self):
    self.EMAIL_REPLY_CONNECTION_MAX_LIFETIME = 'xxx' # str

class UsableConfig(RawConfig):
  def __init__(self):
    pass
```
  @property
  def EMAIL_REPLY_CONNECTION_MAX_LIFETIME(self):
    return int(parent.EMAIL_REPLY_CONNECTION_MAX_LIFETIME)
```

Another design is:

- set on a generic way all values found in ENV variables
- then override specific one by an explicite cast

eg:

```
set_from_env(cfg: CFG):
  for attr in CFG.attr_list:
    if  os.env.get(attr)
        cfg.attr =  os.env.get(attr)
  
  if os.env.get('EMAIL_REPLY_CONNECTION_MAX_LIFETIME'):
    cfg.EMAIL_REPLY_CONNECTION_MAX_LIFETIME = int(os.env.get('EMAIL_REPLY_CONNECTION_MAX_LIFETIME'))
```

**Comment by inkhey** [comment]
Decision has been made to choose code like :-1:
```
def get_raw_config(self, config_name: str, default_value: str) -> str:
    val_cfg = self.config_file.get(config_name, default_value)
    val_name_env = "TRACIM_{}".format(config.name.upper())  # Note : il faut remplacer les . par des _
    val_env = os.env.get(val_name_env)
    if val_env:
      logging.info("PARAM {} = {} from ENV".format(val_name_env,val_env)
      return val_env
    else:
      logging.info("PARAM {} = {} from CONFIG".format(val_name_env,val_env)
      return val_cfg

    self.MY_CONFIG = int(self.get_raw_config("my_config", 42))
    self.MY_OTHER_CONFIG = self.get_raw_config("my_config", "foo,bar").split(",")
```

**Comment by inkhey** [comment]
@PhilippeAccorsi Pour test : 
color.config_file_path --> `TRACIM_COLOR__CONFIG_FILE_PATH=$TRAVIS_BUILD_DIR/backend/color-test.json`
(commence par TRACIM_ , en majuscule '.' devient "__" )

**Comment by PhilippeAccorsi** [comment]
- Some improvement is necessary for some parameter #1672 
- Need possibility to generate list of available parameter with good syntaxe. For the moment is not easy to know which parameter is available or not

**Comment by PhilippeAccorsi** [comment]
For the moment, all available parameters in tracim 2.2 are visible here https://github.com/tracim/tracim/blob/develop/backend/doc/setting.md

**Comment by PhilippeAccorsi** [comment]
Verified.
